### PR TITLE
Allow stdlib freeze

### PR DESCRIPTION
**Pull request by palaviv**
As discussed in #1142.
Adding the current `Lib` folder increase the executable by 5MB. I would like to add this to the demo so it will support python stdlib.
I wanted to allow specifying the location of the stdlib dir with a compilation flag but could not find out how to do that. @coolreader18 any suggestions?

### CODE REVIEW

File: derive/src/compile_bytecode.rs [hunk 1]
```diff
@@ -55,10 +65,61 @@ impl CompilationSource {
...
+            }
+            CompilationSourceKind::SourceCode(code) => {
+                hashmap! {module_name.clone() => self.compile_string(code, mode, module_name.clone())?}
+            }
+            CompilationSourceKind::Dir(rel_path) => {
+                let mut path = PathBuf::from(
+                    env::var_os("CARGO_MANIFEST_DIR").expect("CARGO_MANIFEST_DIR is not present"),
+                );
+                path.push(rel_path);
+                self.compile_dir(&path, "".to_string(), mode)?
```
Comment (coolreader18) [inline derive/src/compile_bytecode.rs]:
```suggestion
                self.compile_dir(&path, String::new(), mode)?
```

File: derive/src/compile_bytecode.rs [hunk 2]
```diff
@@ -154,16 +225,22 @@ impl Parse for PyCompileInput {
...
+        quote! {  #module_name.into() => bincode::deserialize::<::rustpython_vm::bytecode::CodeObject>(#bytes)
+                .expect("Deserializing CodeObject failed") }
+    });
 
     let output = quote! {
         ({
             use ::rustpython_vm::__exports::bincode;
-            bincode::deserialize::<::rustpython_vm::bytecode::CodeObject>(#bytes)
-                .expect("Deserializing CodeObject failed")
+            hashmap! {
```
Comment (coolreader18) [inline derive/src/compile_bytecode.rs]: You should probably add `hashmap` to `__exports` and use it here.

**Comment by windelbouwman** [comment]
Btw, I opened this issue in pyckitup: https://github.com/pickitup247/pyckitup/issues/4 they might be interested in this.

File: derive/src/compile_bytecode.rs [hunk 3]
```diff
@@ -55,10 +65,61 @@ impl CompilationSource {
...
+                );
+                path.push(rel_path);
+                self.compile_dir(&path, "".to_string(), mode)?
+            }
+        })
+    }
+
+    fn compile_dir(
+        &self,
+        path: &PathBuf,
```
[the remark this answers is not in the document]
Comment (palaviv) [reply derive/src/compile_bytecode.rs]: Why is this better?

File: derive/src/compile_bytecode.rs (hunk 2 above)
Comment (palaviv) [reply derive/src/compile_bytecode.rs]: What is the purpose of the `__exports`? This is my first time using procedural macro

Comment (palaviv) [reply derive/src/compile_bytecode.rs]: I think I understand what `__exports` do. Added `hashmap`

**Comment by palaviv** [comment]
> Btw, I opened this issue in pyckitup: [pickitup247/pyckitup#4](https://github.com/pickitup247/pyckitup/issues/4) they might be interested in this.

We should probably add the ability to freeze any directory.


> LGTM. An idea for allowing specifying the stdlib dir, you could maybe allow a `DirFromEnv` `CompilationSourceKind`?

I can get the directory from an environment variable but I hoped for something easier.

**Comment by coolreader18** [comment]
Ah yeah, I think that's the best way to go for configuration, as Rust doesn't have anything like that built in. We've done similar configuration in env variables before, e.g. `BUILDTIME_RUSTPYTHONPATH`

File: derive/src/compile_bytecode.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (coolreader18) [reply derive/src/compile_bytecode.rs]: `PathBuf` vs `Path` is akin to `String` vs `str` - an owned, allocated buffer vs a slice. Like you should (almost) always prefer `&str` over `&String`, so should you for `&Path` over `&PathBuf`.

**Comment by coolreader18** [comment]
I sorta glazed over this when I reviewed earlier, but I don't think that it should always return a `HashMap`, it should only do that when the source is `Dir`.

**Comment by palaviv** [comment]
> I sorta glazed over this when I reviewed earlier, but I don't think that it should always return a `HashMap`, it should only do that when the source is `Dir`.

I am not sure about that... I prefer that `py_compile_bytecode` will return the same type on all cases. I think it will be confusing otherwise.

**Pull request merged**